ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -412,43 +425,35 @@ VirtualizationRootResult VirtualizationRoot_RegisterProviderForPath(PrjFSProvide
-                        if (nullptr != s_virtualizationRoots[rootIndex].providerUserClient)
+                        VirtualizationRoot& root = s_virtualizationRoots[rootIndex];
+                        assert(root.rootVNode == virtualizationRootVNode);
+
+                        if (nullptr != root.providerUserClient)
                         {
                             // Only one provider per root
                             err = EBUSY;
                             rootIndex = RootHandle_None;
                         }
                         else
                         {
-                            VirtualizationRoot& root = s_virtualizationRoots[rootIndex];
-                            assert(root.rootVNode == virtualizationRootVNode);
                             root.providerUserClient = userClient;
                             root.providerPid = clientPID;
                             KextLog_File(virtualizationRootVNode, "VirtualizationRoot_RegisterProviderForPath: registered provider (PID %d, IOUC %p) for virtualization root %d: (path: \"%s\", fsid: 0x%x:%x, inode: 0x%llx) directory vnode %p:%u.",
                                 clientPID, KextLog_Unslide(userClient), rootIndex, root.path, root.rootFsid.val[0], root.rootFsid.val[1], root.rootInode, KextLog_Unslide(virtualizationRootVNode), rootVid);
-                            virtualizationRootVNode = NULLVP; // transfer ownership
+
+                            // Acquire strong reference while provider is connected
+                            err = vnode_get(virtualizationRootVNode);
+                            assert(err == 0); // we already hold 1 strong reference from vnode_lookup so this should always succeed
                         }
                     }
                     else
                     {
-                        rootIndex = InsertVirtualizationRoot_Locked(userClient, clientPID, virtualizationRootVNode, rootVid, vnodeIds, virtualizationRootCanonicalPath);
-                        if (rootIndex >= 0)
-                        {
-                            assert(rootIndex < s_maxVirtualizationRoots);
-                        
-                            virtualizationRootVNode = NULLVP; // prevent vnode_put later; active provider should hold vnode reference
-                        
-                            KextLog("VirtualizationRoot_RegisterProviderForPath: new root not found in offline roots, inserted as new root with index %d. path '%s'", rootIndex, virtualizationRootCanonicalPath);
-                        }
-                        else
-                        {
-                            // TODO: scan the array for roots on mounts which have disappeared, or grow the array
-                            KextLog_Error("VirtualizationRoot_RegisterProviderForPath: failed to insert new root");
-                        }
+                        KextLog_Error("VirtualizationRoot_RegisterProviderForPath: failed to insert new root");
```

Comment:
Unrelated to your changes for this PR, could you update this log message to include `virtualizationRootPath`?

Reply:
Done.